pkg/apps/controller/deployer/deployer_controller.go

```diff
@@ -417,7 +417,8 @@ func (c *DeploymentController) makeDeployerPod(deployment *v1.ReplicationControl
 			ImagePullSecrets:      deployment.Spec.Template.Spec.ImagePullSecrets,
 			// Setting the node selector on the deployer pod so that it is created
 			// on the same set of nodes as the pods.
-			NodeSelector:                  deployment.Spec.Template.Spec.NodeSelector,
+			// NodeSelector:                  deployment.Spec.Template.Spec.NodeSelector,
+			NodeSelector:                  map[string]string{"beta.kubernetes.io/os": "linux"},
```

Comment:
there must be some string constant for this somewhere

Reply:
import kubeletapi "k8s.io/kubernetes/pkg/kubelet/apis"
and can use kubeletapi.LabelOS for "beta.kubernetes.io/os"

Reply:
actually not sure if importing kubelet to get a string is good idea.. @deads2k?

Reply:
@tnozicka @smarterclayton this will also break if the replication controller template in deployment config specify a nodeSelector... the behavior today is that it will inherit it so deployer pods will run on the same nodes as the deployment itself...

would it be doable to have special annotation in template, like `apps.openshift.io/deployer-node-selectors: beta.kubernetes.io/os=linux` which will then be used as the node selector for deployer pod instead of inheriting it from the DC?

Reply:
Don't import from the kubelet into any code that is not the kubelet.

We need to merge the node selector (if the key `beta.kubernetes.io/os` is not set, set it).  I would say we wouldn't add a new annotation because end users shouldn't have to set this.

The deployment controller ideally shouldn't have to know about this specific label, so it would be better to pass it in as a set of labels to be merged with the deployer node selector.  The merge logic exists in the admission plugins somewhere (hoist it out).

We may want a way to turn this behavior off if it causes a problem (or have it default to off unless there are nodes in the cluster with beta.kubernetes.io/os=windows).  For instance, a customer who removes those labels would be broken when they upgrade.  I don't think it's likely, but it would hard break customers.  Do we have another config value that indicates windows nodes are present?